### ISSUE TITLE
What does cudaGetDeviceCount() and cudaMallocManaged() unknown error mean?

### ISSUE THREAD

**Issue by bartlettroscoe** [kokkos/kokkos#2377]
Hello Kokkos developers,

In some of that ATDM Trilinos CUDA builds from time-to-time we see mass test failures showing errors like:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaGetDeviceCount( & m_cudaDevCount ) error( cudaErrorUnknown): unknown error /home/jenkins/waterman/workspace/Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug/SRC_AND_BUILD/Trilinos/packages/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:204
Traceback functionality not available

[waterman1:15443] *** Process received signal ***
[waterman1:15443] Signal: Aborted (6)
[waterman1:15443] Signal code:  (-6)
...
```

like shown in [this test](https://testing-dev.sandia.gov/cdash/testDetails.php?test=75738238&build=4883796) as of of the [1011 failing tests in the build Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug on 2019-09-25 showing this error](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&date=2019-09-25&filtercount=5&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug&field2=status&compare2=62&value2=passed&field3=status&compare3=62&value3=notrun&field4=testoutput&compare4=95&value4=cudaGetDeviceCount&field5=testoutput&compare5=95&value5=unknown%20error) and:

```
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name waterman1 and rank 0!
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaMallocManaged( (void**) &m, sz*sizeof(T), cudaMemAttachGlobal ) error( cudaErrorUnknown): unknown error /home/jenkins/waterman/workspace/Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug/SRC_AND_BUILD/Trilinos/packages/sacado/src/Sacado_DynamicArrayTraits.hpp:195
Traceback functionality not available

[waterman1:30399] *** Process received signal ***
[waterman1:30399] Signal: Aborted (6)
[waterman1:30399] Signal code:  (-6)
```

for [this test](https://testing-dev.sandia.gov/cdash/testDetails.php?test=75738524&build=4883808)  as one of the [29 filing tests in the build Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug on 2019-09-25 showing this error](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&date=2019-09-25&filtercount=5&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug&field2=status&compare2=62&value2=passed&field3=status&compare3=62&value3=notrun&field4=testoutput&compare4=95&value4=cudaMallocManaged&field5=testoutput&compare5=95&value5=unknown%20error).

Anyone seen situations like this?  Does this mean the GPU is being overloaded?   If so, it seems to be occurring somewhat randomly and not just in RDC builds as shown in [this query](https://testing-dev.sandia.gov/cdash/index.php?project=Trilinos&begin=2019-09-01&end=now&filtercount=1&showfilters=1&field1=buildname&compare1=65&value1=Trilinos-atdm-waterman-cuda-).  Could this mean there is a bad GPU or something on 'waterman'?

Not expecting Kokkos developers to triage this.  Just asking if you have seen this before and might know what causes this.

**Comment by stanmoore1** [comment]
I see this too.

>Could this mean there is a bad GPU or something on 'waterman'?

Yes

**Comment by bartlettroscoe** [comment]
@stanmoore1 said:

> I see this too.

Should we be looking to if there is particular node 'waterman' where this occurs?  In the past, I have found bad nodes on 'ride' that we had to exclude from automated testing.

BTW: The internal tracker for this is:

* https://sems-atlassian-srn.sandia.gov/browse/ATDV-205

**Comment by stanmoore1** [comment]
>Should we be looking to if there is particular node 'waterman' where this occurs?

Yes, it could as easy as running `nvidia-smi` on each node to see if there are 4 GPUs registered.

**Comment by bartlettroscoe** [comment]
FYI: Looks like all of the instances where we see these mass test failures occur when running on the node 'waterman1' but not always when running on there.

I will exclude 'waterman1' from being used in our automated testing and see if that fixes the problem.

Thanks!

**Comment by crtrott** [comment]
bad gnu driver or something like that.
